### PR TITLE
NEW set draft shipping

### PR DESCRIPTION
**Pull request by atm-quentin**
# New Set draft shipping

Allow people to set shipping status to draft

### CODE REVIEW

**Comment by dolibarr95** [comment]
Hi !
See https://github.com/Dolibarr/dolibarr/pull/8953

File: htdocs/expedition/class/expedition.class.php [hunk 1]
```diff
@@ -2257,6 +2256,138 @@ function reOpen()
 		}
 	}
 
+	// phpcs:disable PEAR.NamingConventions.ValidFunctionName.NotCamelCaps
```
Comment (frederic34) [inline htdocs/expedition/class/expedition.class.php:2295]: phpcs rule was introduced to avoid adding non camel caps function

Comment (atm-quentin) [reply htdocs/expedition/class/expedition.class.php:2295]: Hello, yes i understand but i wanted to keep the format of this function like in other object see : https://github.com/Dolibarr/dolibarr/blob/8ee98315c750b2fdd24c055dc6a1cd2d7a1ffee4/htdocs/commande/class/commande.class.php#L461

Comment (frederic34) [reply htdocs/expedition/class/expedition.class.php:2295]: so don't forget the phpcs:enable

**Comment by eldy** [comment]
What is use case of setting back to draft a shipment (user has clicked by error ?).
Because when a shipment was validated, the stock were also moved.

**Comment by atm-quentin** [comment]
For some customer, the truck driver come at midnight (when there's noone in company) and need to validate shipment to print it. But after that sometimes they don't take exactly the same quantity, so they need to set it back to draft and change the amount the day after to correct the stock. They can't know exactly the quantity the truck driver will take before he comes.

In the code i canceled the stock too.

**Comment by frederic34** [comment]
On a shipping with many many lines, it will be great to be able to return to draft just to correct one line.

**Comment by atm-maxime** [comment]
I think many users use shipment to print the packing list and go prepare the shipment. And when the stock is incorrect, they would like to modify the shipment to enter the correct quantities, and also correct stock... Going back to draft does that.

**Comment by BebZ** [comment]
The remark of Maxime is true, setting back to draft shipment is necessary to correct one or more lines with stock adjustments off course. Is this function possible yet ? I mean correct a shipment and that makes stock movements ?

**Comment by atm-quentin** [comment]
Is this function possible yet ? I mean correct a shipment and that makes stock movements ? ==> Actually not, that's why i made this pull request

**Comment by atm-quentin** [comment]
@eldy What do you think about it ?

**Comment by atm-maxime** [comment]
Ping @eldy

**Comment by atm-quentin** [comment]
Ping @eldy

**Comment by dolibarr95** [comment]
:-p 
See #8953

**Comment by eldy** [comment]
To analyze also with #10174 #8953

**Comment by eldy** [comment]
If there is something wrong after a shipment was done (so stock movement was recorded), and we need to send more, we should not go back to draft and resent the shipment. This has a lot of side effect.
There is a lot of different cases:
1) The shipment was a picking list with quantity X and we send Y, stock was decrease of X. It means the transporter has made an error by not following the picking list. In such cas we must have a feature to make a stock correction from the shipment (the stock movement will also be linked to the shipment like the movement done automatically).
2) The shipment was a draft picking list with quantity X and we send X, but there was an error, we would like to send X+1. Because we have to send +1 again we have to make another shipment. It is a second sending, so a second shipment that can be linked to first one.
3) The shipment was a draft picking list with quantity X and we send X but customer return Y, we must be able to generate what is called a "back order". It should be an action button on shipment that should be available from the shipment too that make the stock movement linked to the shipment.

The trouble is that the shipment is "validated" when it should not. "Validate" a shipment means, the carrier has leaved the warehouse and the stock has been definitely decreased. I think the trouble is that we miss a status between "Draft" and current "Validate". The current "Validate" status should be renamed into "Sent" and the new status should be called "Validate". Then changing between Draft and new Validate will not be a problem.
This point has also been tracked as a feature request in the last point not yet done of Issue #4365

Another solution to adresse that is to add option "Decrease real stocks on shipping classification to Closed / Shipped". When this option is on, you must be able to enter the batches before going to closed/shipped ?
Make it possible to modify the number shipped and lot when the shipment is in "Validated" status.

**Comment by torvista** [comment]
Note
"Back order": when a customer makes an order but the supplier does not have enough stock, the items that cannot be supplied are on "back order" at the supplier.
So, customer orders X, but only Y can be sent, means (X-Y) is on "back-order".
That does not seem to match your use of the term.

From the english language point of view, I would use
Draft
Dispatch
Validated

File: htdocs/expedition/class/expedition.class.php (hunk 1 above)
Comment (frederic34) [reply htdocs/expedition/class/expedition.class.php:2295]: https://github.com/Dolibarr/dolibarr/commit/c0b6699b253dbcc8aeba51bc90e3d4f1a9323e32
all methods have been renamed to setDraft

**Comment by atm-maxime** [comment]
Hi @eldy. I agree with the fact that the statuses and workflows must change. I think this could be done simply by changing the closed status to "sent" (and adding a billed field yes/no as on orders). Then when the stock decrease option is set on shipping close, there's no problem allowing a validated shipment to go back on draft status to allow quantity change. This last thing can be done very easily with a test on stock decrease option. What do you think ?

**Comment by atm-maxime** [comment]
@eldy ?

**Comment by eldy** [comment]
There is several PR to enhance this and it is not clear what will be the behaviour after (except that all of them suggest a different way of working). So to help to choose which one to retain, i updated wiki doc to describe current worklow:
https://wiki.dolibarr.org/index.php/Module_Shipments#Usage

Can you cut and paste this workflow text graph and update it to describe what will be workflow after the change (keeping the X and the number of status in parenthesis)

**Comment by atm-quentin** [comment]
Draft  ---X---> Validated ---X---> Draft 
It just allows you to set back your shipment to draft.
If stock is updating on validation of shipment, we create a stock movement in order to cancel it

**Comment by eldy** [comment]
I don't think going back to draft by making a revert movement is a good solution. I think we should enhance the process with something like that:
Draft ---X---> Validated ---> Closed    (X = stock change, Can edit qty and lot in Draft step)
must become
Draft ----> Validated ---X---> Shipped   (X = stock change, Can edit qty and lot in status Draft + Validated)

**Pull request closed without merging**